numba/annotations/type_annotations.py

```diff
@@ -0,0 +1,271 @@
+from __future__ import print_function, absolute_import
+import inspect
+import re
+from collections import Mapping, defaultdict
+import textwrap
+from contextlib import closing
+from numba.io_support import StringIO
+from numba import ir
+import numba.dispatcher
+import os
+import sys
+
+
+class SourceLines(Mapping):
+    def __init__(self, func):
+
+        try:
+            lines, startno = inspect.getsourcelines(func)
+        except IOError:
+            self.lines = ()
+            self.startno = 0
+        else:
+            self.lines = textwrap.dedent(''.join(lines)).splitlines()
+            self.startno = startno
+
+    def __getitem__(self, lineno):
+        try:
+            return self.lines[lineno - self.startno].rstrip()
+        except IndexError:
+            return ''
+
+    def __iter__(self):
+        return iter((self.startno + i) for i in range(len(self.lines)))
+
+    def __len__(self):
+        return len(self.lines)
+
+    @property
+    def avail(self):
+        return bool(self.lines)
+
+
+class TypeAnnotation(object):
+    
+    # func_data dict stores fancy annotation data for all functions that are
+    # compiled. We store the data in the TypeAnnotation class since a new
+    # TypeAnnotation instance is created for each function that is compiled.
+    # For every function that is compiled, we add the type annotation data to
+    # this dict and write the html annotation file to disk (rewrite the html
+    # file for every function since we don't know if this is the last function
+    # to be compiled).
+    func_data = {}
+
+    def __init__(self, interp, typemap, calltypes, lifted, lifted_from, args, return_type,
+                 func_attr, fancy=None):
+        self.filename = interp.bytecode.filename
+        self.func = interp.bytecode.func
+        self.blocks = interp.blocks
+        self.typemap = typemap
+        self.calltypes = calltypes
+        if fancy is None:
+            self.fancy = None
+        else:
+            self.fancy = os.path.join(os.getcwd(), fancy)
+        self.filename = interp.loc.filename
+        self.linenum = str(interp.loc.line)
+        self.signature = str(args) + ' -> ' + str(return_type)
+        self.func_attr = func_attr
+        
+        # lifted loop information
+        self.lifted = lifted
+        self.num_lifted_loops = len(lifted)
+
+        # If this is a lifted loop function that is being compiled, lifted_from
+        # points to annotation data from function that this loop lifted function
+        # was lifted from. This is used to stick lifted loop annotations back
+        # into original function.
+        self.lifted_from = lifted_from
+
+    def prepare_annotations(self):
+        # Prepare annotations
+        groupedinst = defaultdict(list)
+        found_lifted_loop = False
+        #for blkid, blk in self.blocks.items():
```

Comment:
I think this would work, assuming you add a sorted() call around `self.block.items()`.